crates/ruff_server/src/server/api/traits.rs

```diff
@@ -0,0 +1,142 @@
+//! A stateful LSP implementation that calls into the Ruff API.
+
+use crate::server::client::{Notifier, Responder};
+use crate::session::{Session, SessionSnapshot};
+
+use lsp_server as server;
+use lsp_types::notification::Notification as LSPNotification;
+use lsp_types::request::Request as LSPRequest;
+
+/// A supertrait for any server request handler.
+pub(super) trait Request {
+    type RequestType: LSPRequest;
+    const METHOD: &'static str = <<Self as Request>::RequestType as LSPRequest>::METHOD;
+
+    /// Tries to cast a serialized request from the server into
+    /// a parameter type for a specific request handler.
+    /// It is *highly* recommended to not override this function in your
+    /// implementation.
```

Comment:
Can you add more to these docs explaining why?

Reply:
I actually might move the methods on this trait and `Notification` out into standalone functions as part of the work to remove `select_task!`